File: leetcode/problems/language-specific/go/121.best-time-to-buy-and-sell-stock.go
/*
 * @lc app=leetcode id=121 lang=golang
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
func maxProfit(prices []int) int {
	if len(prices) <= 1 {
		return 0
	}

	left, maxProfit := 0, 0
	for right := 1; right < len(prices); right++ {
		if prices[right] > prices[left] {
			maxProfit = findMax(maxProfit, prices[right]-prices[left])
		} else {
			left = right
		}
	}

	return maxProfit
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/language-specific/python/121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start

# * Two Pointers and Sliding Window Solution | O(n) Time | O(1) Space
# * n -> The length of prices array


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        prices_len = len(prices)
        max_profit = 0
        left = 0
        for right in range(1, prices_len):
            if prices[right] > prices[left]:
                max_profit = max(max_profit, prices[right] - prices[left])
            else:
                left = right

        return max_profit


# @lc code=end
